### PR TITLE
Shared ptr + weak ptr

### PR DESCRIPTION
**Pull request by MonikaZelechowska**

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Good job :+1:

File: exercises/uniquPointer.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ class testClass
+{
+public:
+    bool printingFunction()
+    { 
+        return true;
+    }
+};
+
+void testStarOperator()
```
Comment (J-Cieplinski) [inline exercises/uniquPointer.cpp:111]: Why all of those test functions here?

File: exercises/uniquPointer.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    uniqPtrType* operator->();
+    uniqPtrType* get()const;
+    uniqPtrType* release();
+    void reset(uniqPtrType* object);
+};
+
+template <class uniqPtrType>
+UniquePtr<uniqPtrType>::UniquePtr()
+    :pointer(nullptr)
+    {std::cout<<"konstruktor"<<std::endl;}
```
Comment (J-Cieplinski) [inline exercises/uniquPointer.cpp:29]: Is that necessary here?

File: exercises/uniquPointer.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        pointer = object.pointer;
+        object.pointer = nullptr;
+    }
+    return *this;
+}
+
+template <class uniqPtrType>
+uniqPtrType& UniquePtr<uniqPtrType>::operator*() 
+{
+    return *(this->pointer);
```
Comment (J-Cieplinski) [inline exercises/uniquPointer.cpp:62]: dereferencing a nullptr

File: exercises/List.cpp [hunk 4]
```diff
@@ -74,21 +84,83 @@ Node* List::get(const int value)
...
+    {
+        if(temp->value == number)
+        {
+            std::cout<<"szukana wartosc: "<<temp->value<<std::endl;
+            return temp;
+        }
+        else
+        {
+            temp = temp->prev;
+        }
```
Comment (WBigaj) [inline exercises/List.cpp:120]:
```suggestion
        temp = temp->prev;
```

Comment (WBigaj) [reply exercises/List.cpp:120]: This else is redundant since return would escape function.

File: exercises/uniquPointer.cpp (hunk 2 above)
Comment (WBigaj) [reply exercises/uniquPointer.cpp:29]: +1

File: exercises/uniquPointer.cpp (hunk 1 above)
Comment (WBigaj) [reply exercises/uniquPointer.cpp:111]: I agree with @J-Cieplinski. Tests should be in antoher file.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: homework/shared_ptr/WeakPtr.hpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        ptr_ = otherWeakPtr.ptr_;
+        controlBlock_ = otherWeakPtr.controlBlock_;
+    }
+    return *this;
+}
+
+template <typename T>
+WeakPtr<T>& WeakPtr<T>::operator=(const SharedPtr<T>& sharedPtr) noexcept {
+    if (controlBlock_ != nullptr) {
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr.hpp:84]: what about calling `handleWeakPtrAndControlBlockDelete();`? We may already own an object.

File: homework/shared_ptr/WeakPtr.hpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+template <typename T>
+WeakPtr<T>::WeakPtr(WeakPtr&& otherWeakPtr) noexcept
+    : ptr_(std::move(otherWeakPtr.ptr_)), controlBlock_(std::move(otherWeakPtr.controlBlock_)) {
+    otherWeakPtr.ptr_ = nullptr;
+    otherWeakPtr.controlBlock_ = nullptr;
+}
+
+template <typename T>
+WeakPtr<T>& WeakPtr<T>::operator=(const WeakPtr& otherWeakPtr) noexcept {
+    if (&otherWeakPtr != this) {
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr.hpp:72]: what about calling `handleWeakPtrAndControlBlockDelete();`? we may already own an object.

File: homework/shared_ptr/WeakPtr-ut.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    WeakPtrTest()
+        : testWeakPtr(testSharedPtr) {}
+
+    SharedPtr<int> testSharedPtr{new int{initValue}};
+    WeakPtr<int> testWeakPtr;
+};
+
+TEST_F(WeakPtrTest, ShouldGetOneUseCount) {
+    ASSERT_EQ(testWeakPtr.use_count(), useCounter);
+    ASSERT_EQ(testWeakPtr.use_count(), testSharedPtr.use_count());
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr-ut.cpp:19]: you should also check if use_count on weak_ptr drops to zero after resetting shared_ptr and epxired returns true

File: homework/shared_ptr/WeakPtr-ut.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+TEST_F(WeakPtrTest, ShouldResetPtr) {
+    testWeakPtr.reset();
+    ASSERT_TRUE(testWeakPtr.expired());
+}
+
+TEST_F(WeakPtrTest, ShouldPtrBeCreatedByCopyCtor) {
+    WeakPtr<int> copyTestWeakPtr(testWeakPtr);
+
+    ASSERT_FALSE(copyTestWeakPtr.expired());
+    ASSERT_FALSE(testWeakPtr.expired());
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr-ut.cpp:40]: you can check if both of them points to the same data. The same for below test.

File: homework/shared_ptr/WeakPtr-ut.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    ASSERT_EQ(*lockSharedPtr, initValue);
+    ASSERT_FALSE(testWeakPtr.expired());
+}
+
+TEST_F(WeakPtrTest, ShouldEmptyTestWeakPtrLockReturnEmptySharedPtr) {
+    WeakPtr<int> emptyWeakPtr{};
+    SharedPtr<int> lockSharedPtr{emptyWeakPtr.lock()};
+
+    ASSERT_EQ(lockSharedPtr.get(), nullptr);
+    ASSERT_FALSE(testWeakPtr.expired());
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr-ut.cpp:78]: we should test emptyWeakPtr here and it should be true.

**Review by ziobron — CHANGES_REQUESTED** [review]
There are problems with memory that are reported by valgrind:
https://github.com/MonikaZelechowska/memory_management/runs/1238041615?check_suite_focus=true

Apart from that very clean code and tests 👍

File: homework/shared_ptr/SharedControlBlock.hpp [hunk 10]
```diff
@@ -18,9 +18,47 @@ class SharedControlBlock {
...
+    ~SharedControlBlockObj() { defaultDeleter(obj_); }
+
+    void callDefaultDeleter() {
+        defaultDeleter(obj_);
+    }
+    T* getObj() override { return obj_; };
+
+private:
+    T* obj_ = nullptr;
+    std::function<void(T*)> defaultDeleter = [](T* ptr = nullptr) { delete ptr; };
```
Comment (ziobron) [inline homework/shared_ptr/SharedControlBlock.hpp:50]: this hides members from a base class. You should rather remove it from a base class.

File: homework/shared_ptr/SharedPtr.hpp [hunk 11]
```diff
@@ -20,55 +26,72 @@ class SharedPtr {
...
 
-    void handleSharedPtrAndControlBlockDelete();
-
 private:
     T* ptr_ = nullptr;
     SharedControlBlock<T>* shControlBlock_ = nullptr;
 
+    void handleSharedPtrAndControlBlockDelete();
+    explicit SharedPtr(SharedControlBlockData<T>* newBlock)
+        : shControlBlock_(newBlock) { ptr_ = newBlock->getObj(); }
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:43]: do everything on init list

File: homework/shared_ptr/SharedControlBlock.hpp [hunk 12]
```diff
@@ -18,9 +18,47 @@ class SharedControlBlock {
...
+};
+
+template <typename T>
+class SharedControlBlockData : public SharedControlBlock<T> {
+public:
+    template <typename... Args>
+    SharedControlBlockData(Args... args)
+        : object_(args...) {}
```
Comment (ziobron) [inline homework/shared_ptr/SharedControlBlock.hpp:58]: also, a deleter should be overridden to not delete anything.

File: homework/shared_ptr/SharedPtr.hpp [hunk 13]
```diff
@@ -20,55 +26,72 @@ class SharedPtr {
...
-    : ptr_(ptr), shControlBlock_(new SharedControlBlock<T>{deleter}) {}
+    : ptr_(ptr), shControlBlock_(new SharedControlBlockObj<T>{ptr, deleter}) {}
+
+template <typename T>
+SharedPtr<T>::SharedPtr(const WeakPtr<T>& weakPtr)
+    : ptr_(weakPtr.ptr_), shControlBlock_(weakPtr.controlBlock_) {
+    if (shControlBlock_ != nullptr) {
+        shControlBlock_->incrementSharedRefsCount();
+    }
+}
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:79]: what if weak_ptr is expired? we should be protected against it.

File: homework/shared_ptr/MakeShared-ut.cpp [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    cs::SharedPtr<int> shPtr(new int{intTestVal});
+    auto makeShPtr = cs::makeShared<int>(intTestVal);
+    ASSERT_EQ(*shPtr, *makeShPtr);
+}
+
+TEST(MakeSharedPtrTest, ShouldMakeSharedPtrToTestStruct) {
+    auto makeShPtr = cs::makeShared<testStruct>(intTestVal, doubleTestVal);
+    ASSERT_EQ(makeShPtr->intValue_, intTestVal);
+    ASSERT_EQ(makeShPtr->doubleValue_, doubleTestVal);
+}
```
Comment (ziobron) [inline homework/shared_ptr/MakeShared-ut.cpp:27]: please add tests for copying and moving shared_ptrs created by make_unique. They handle resources in a different way.

**Review by ziobron — CHANGES_REQUESTED** [review]

**Review by ziobron — APPROVED** [review]
Well done :)

**Comment by ziobron** [comment]
50 XP za implementację shared_ptra (3794)
25 XP za implementację weak_ptra (3795)
40 XP za implementację make_shared (3796)
{ "requirement_ids": [3794, 3795, 3796] }

**Comment by ziobron** [comment]
{
"requirement_ids": [1119, 1119, 1119]
}

**Pull request closed without merging**